File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/rule/LuhnRule.java
package com.wesleyelliott.kubwa.rule;

import android.text.TextUtils;

/**
 * Created by wesley on 2016/08/04.
 */

public abstract class LuhnRule extends Rule<String> {

    private int modulus;

    public LuhnRule(int modulus) {
        super(String.class);
        this.modulus = modulus;
    }

    public boolean validate(String value) {
        return validate(value, ".+", true);
    }

    public boolean validate(String value, String regex) {
        return validate(value, regex, true);
    }

    public boolean validate(String value, String regex, boolean includesCheckDigit) {
        // Validate Regex
        if (TextUtils.isEmpty(value) || !value.matches(regex)) {
            return false;
        }

        int total = 0;
        for (int i = 0; i < value.length(); i++) {
            int lth = value.length() + (includesCheckDigit ? 0 : 1);
            int leftPos  = i + 1;
            int rightPos = lth - i;
            int charValue = toInt(value.charAt(i));
            total += weightedValue(charValue, leftPos, rightPos);
        }
        if (total == 0) {
            return false;
        }
        return (total % modulus) == 0;
    }

    private int weightedValue(int charValue, int leftPos, int rightPos) {
        int[] POSITION_WEIGHT = new int[] {2, 1};
        int weight = POSITION_WEIGHT[rightPos % 2];
        int weightedValue = charValue * weight;
        return weightedValue > 9 ? (weightedValue - 9) : weightedValue;
    }

    private int toInt(char character) {
        if (Character.isDigit(character)) {
            return Character.getNumericValue(character);
        }
        return -1;
    }

}

File: kubwa-annotations/src/main/java/com/wesleyelliott/kubwa/rule/LuhnRule.kt
package com.wesleyelliott.kubwa.rule

import com.wesleyelliott.kubwa.Utils

/**
 * Created by wesley on 2016/08/04.
 */

abstract class LuhnRule(private val modulus: Int) : Rule<String>(String::class.java) {

    @JvmOverloads
    fun validate(value: String, regex: String = ".+", includesCheckDigit: Boolean = true): Boolean {
        // Validate Regex
        if (Utils.isEmpty(value) || !value.matches(regex.toRegex())) {
            return false
        }

        var total = 0
        for (i in value.indices) {
            val lth = value.length + if (includesCheckDigit) 0 else 1
            val leftPos = i + 1
            val rightPos = lth - i
            val charValue = toInt(value[i])
            total += weightedValue(charValue, leftPos, rightPos)
        }
        return if (total == 0) {
            false
        } else total % modulus == 0
    }

    private fun weightedValue(charValue: Int, leftPos: Int, rightPos: Int): Int {
        val POSITION_WEIGHT = intArrayOf(2, 1)
        val weight = POSITION_WEIGHT[rightPos % 2]
        val weightedValue = charValue * weight
        return if (weightedValue > 9) weightedValue - 9 else weightedValue
    }

    private fun toInt(character: Char): Int {
        return if (Character.isDigit(character)) {
            Character.getNumericValue(character)
        } else -1
    }

}
